### PR TITLE
Add detailed documentation

### PR DESCRIPTION
**Pull request by rgantonio**
This PR adds documentation on how to use several tools in this repo. Changes in this PR:

* Took out old container README and moved the guide to the `util` directory.
* Added documentation on how to run tests
* Added documentation on how to use the utility directory
* Added more documentation on the main README, including the port connections for clarity.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,11 +43,81 @@ To hook the pre-commit hooks into git, use:
...
+    * The Cocotb tests function as the main test bench and stimuli for the design under test (DUT).
+    * The test bench wrappers function as top-level wrappers to connect to cocotb.
+    * The `snax_util.py` contains utility components for tests and read-write register or direct memory access (DMA) operations.
+ * `/util` - contains configuration files, scripts, templates, and the container for quick setups.
+
+# Testbench Guide
+
+Below shows a figure for the test and test bench setup:
+
+<img src="https://drive.google.com/thumbnail?id=1fI5fB6a7tJELptoN1RR7dBMUnbwpcmYO&sz=w1960" alt="SNAX Development Setup">
```
Comment (jorendumoulin) [inline README.md:63]: this diagram is super good 🤩

File: util/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+* `/scripts` - Contains useful scripts. Right now we have a template generation tool.
+* `/cfg` - Has configuration files for templates used for the template generation.
+* `/templates` - Contains several templates for different files.
+
+# Docker Usage
+
+You can build the docker with:
+
+```bash
+docker build -t ghcr.io/kuleuven-micas/snax-cocotb .
```
Comment (jorendumoulin) [inline util/README.md:15]: isn't it `docker build util/container -t snax-cocotb`?

File: util/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+This directory has the following:
+
+* `/container` - Has the docker file used for this repo.
+* `/scripts` - Contains useful scripts. Right now we have a template generation tool.
+* `/cfg` - Has configuration files for templates used for the template generation.
+* `/templates` - Contains several templates for different files.
+
+# Docker Usage
+
+You can build the docker with:
```
Comment (jorendumoulin) [inline util/README.md:12]: from scratch

File: util/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+```bash
+docker build -t ghcr.io/kuleuven-micas/snax-cocotb .
+```
+
+You can run the container with (make sure you are at the root of the repo):
```
Comment (jorendumoulin) [inline util/README.md:18]: the prebuilt container

File: README.md [hunk 5]
```diff
@@ -26,11 +43,81 @@ To hook the pre-commit hooks into git, use:
...
+| valid       | Valid signal. High when data is valid.              |
+| ready       | Ready signal. High when receiving end is valid.     |
+
+**Note** Valid-ready response have very specific rules which you can find from the [AXI ARM Documentation](https://developer.arm.com/documentation/102202/0300/Channel-transfers-and-transactions). In summary:
+
+* A successful transaction only happens when both `VALID` and `READY` signals are high.
+* A source cannot wait for READY to be asserted before asserting `VALID`.
+* A destination can wait for VALID to be asserted before asserting `READY`.
+
+The accelerator ports for getting data from streamers uses the same mechanism. They only have one channel which is the `data` port. The interaction between streamers and the accelerator involves only data transfers. If you are to use streamers, refer to the [SNAX Streamer](https://github.com/KULeuven-MICAS/snax-streamer) repository.
```
Comment (jorendumoulin) [inline README.md:99]:
```suggestion
The accelerator ports for getting data from streamers use the same mechanism. They only have one channel which is the `data` port. The interaction between streamers and the accelerator involves only data transfers. If you are to use streamers, refer to the [SNAX Streamer](https://github.com/KULeuven-MICAS/snax-streamer) repository.
```

File: README.md [hunk 6]
```diff
@@ -26,11 +43,81 @@ To hook the pre-commit hooks into git, use:
...
+# Testbench Guide
+
+Below shows a figure for the test and test bench setup:
+
+<img src="https://drive.google.com/thumbnail?id=1fI5fB6a7tJELptoN1RR7dBMUnbwpcmYO&sz=w1960" alt="SNAX Development Setup">
+
+There are three major key sections:
+
+* `TCDM Subsystem` - is at the top-most area. This is a replica of the Snitch's TCDM subsystem where it contains narrow and wide TCDM ports. The narrow are for the accelerator's ports while the wide are for the DMA's ports.
+* `Accelerator Wrapper` - is the main accelerator DUT. The figure shows that it has streamers connected to the main accelerator data path. There also exists a CSR Control which is also a manager for setting configurations. **Note** It is possible to design or have your own streamer, if you have an accelerator built this way, then we connect it directly to the TCDM interconnect following the given ports.
```
Comment (jorendumoulin) [inline README.md:68]:
```suggestion
* `Accelerator Wrapper` - is the main accelerator DUT. The figure shows that it has streamers connected to the main accelerator data path. The CSR Control is a manager for setting accelerator configurations. **Note** It is possible to design or have your own streamer, if you have an accelerator built this way, then we connect it directly to the TCDM interconnect following the given ports.
```

**Review by jorendumoulin — APPROVED** [review]
Nice! Only the docker build commands seem to be a bit off

File: util/README.md (hunk 2 above)
Comment (rgantonio) [reply util/README.md:15]: Good point haha too much copy-pasting hahaha

File: util/README.md (hunk 3 above)
Comment (rgantonio) [reply util/README.md:12]: Fixed!

File: util/README.md (hunk 2 above)
Comment (rgantonio) [reply util/README.md]: Fixed!

File: README.md (hunk 5 above)
Comment (rgantonio) [reply README.md]: Fixed!

File: README.md (hunk 6 above)
Comment (rgantonio) [reply README.md]: Fixed!

**Comment by rgantonio** [comment]
Maraming salamat @jorendumoulin !!! ❤️ ❤️ ❤️ ❤️ ❤️

**Comment by rgantonio** [comment]
Also @jorendumoulin thanks for the suggestions and checks 😆

**Comment by JosseVanDelm** [comment]
Looking good @rgantonio!

**Comment by rgantonio** [comment]
@JosseVanDelm thanks! Sorry I didn't wait for you guys haha

and ok ok! Got it 😄 I will add later in the next READMEs.

**Pull request merged**